### ISSUE TITLE
Abandoned deposit exploit

### ISSUE THREAD

**Issue by Agoristen** [keep-network/tbtc#738]
The attacker, hereby referred to as "abandoner" will exploit the system to increase his own rewards. His attack will be targeted against innocent node operators, hereby referred to as "victims". 

This exploit is currently highly incentivized due to the way keep rewards is to be paid. Rewards are paid based on # of successful deposit + redeems performed within a given period (30 days). Currently, the abondoner will use this exploit to increase his own reward count and therefore rewards. 

**It works like this:**
When a deposit is initiated it locks up bonds for 3 signers. By firing deposits in rapid succession the abandoner is able to force other nodes out of capacity, essentially rendering himself the sole node operator for his chosen lot size. This has centralization risk since all new deposits now will be routed through his own nodes, but furthermore it has a negative impact on the reward schedule as one single actor is able to acquire a higher percent of the rewards than is expected based on assigned KEEP and ETH bond.

While we are yet to see a large scale attack based on this exploit. There are been smaller attempts from less crafty adversaries. 

Here is an example from today:
![Example of abandoned deposits](https://i.imgur.com/RqkzWss.png)

These are all deposits initiated by the same user and are all subsequently abandend. 

If the deposit is abandoned, one of the three victims (signers) will have to run `notifyFundingTimedOut` to get their funds back. However, if they run `notifyFundingTimedOut` before `retrieveSignerPubkey` is called they will be slashed for approximately 0.05 ETH each (~0.144 ETH / 3). The abandoner is not incentivized to run `retrieveSignerPubkey`, in fact he is rewarded for not doing so, as it means the 0.144 ETH is returned to him on notify. 

The victims must perform the following tasks for each abandoned deposit:
1. In the `Awaiting Funding` state they need to call retrieveSignerPubkey manually on the contract
2. In the `Awaiting Funding Proof` state they need to call notifyFundingTimedOut

Currently, this problem is not widespread knowledge, which causes victims to call notify without first calling `retrieveSignerPubkey`, causing themselves (and their co-signers) a slashing. But more stressing is the fact that abandoner is highly incentivized to call Notify after 3 hours to get his full amount returned and slash the victims, Victims therefore has to continually monitor and call `retrieveSignerPubkey` before the abandoner is able to do so. In practice, they will have to call this _before_ it's been 3 hours to protect themselves against slashing. 

Currently, an attack of this type costs roughly 0.06 ETH in txn fee, and each abandoned deposit locks up 150% of lot size between 3 signers. At current capacity, you can lock out the majority of 10 BTC lot nodes with 10-20 abandoned deposits, a negligible cost compared to the advantage gained of a successfully deployed attack. To date there has been only 190 timeouts on the network, however it is expected that the problem will increase as this becomes understood to be an advantage. In fact, timeouts today alone accounts for 25% of all timeouts (total 47) - a clear indication that abandoning is becoming a more rampant issue in the system. 

This exploit could be mitigated by requiring depositor to put up a small bond which will be returned after successful deposit. For example 0.1 ETH that will be distributed to the notifier (a person calling both `retrieveSignerPubkey` and `notifyFundingTimedOut`). A notifier would therefore become an incentivized job on the network and would help clearing the timeout deposits in a timely fashion and opening up capacity in the network for legitimate depositors all while simultaneously preventing a successful large scale attack to be deployed. 

An alternative approach could be to set a higher bond, such as 0.15-0.25 ETH. If deposit is abandonend, this amount will be distributed among the signers on notifyFundingTimedOut to cover their fees and reduce the incentive for abandoning. However I think the current scheme where 1 signer has to take the hit is not a good way to go about it. Doing this work should be rewarded, or else it will not be performed unless absolutely necessary. The result of not rewarding notify timeout is less available capacity in the network. 

To be clear, I not advocating for changing of the reward schedule. But I would like for this issue to be better explored and documented for users in the UI and tools first and foremost. It should further be considered if this can be mitigated by introducing a deposit bond. In addition to a bond, the issue will be mitigated to some extent by having more nodes and more capital in the system. Reducing lot size to max 1 BTC would accomplish the same effect as growing the network, but the disadvantages due to higher relative minting costs makes this a no-go in my view. 

From the victims perspective. The only way to fight this exploit today is to increase ETH bond and constantly monitor and notify the abandonend deposits, resulting in higher risk and/or real costs to victims. 

Some may argue that this is only a problem during the initial 24 months of reward emission, however with the costs being relatively low, I could see this be exploit be used for attacks of more sinister nature than we currently see from reward seekers.

**Comment by mhluongo** [comment]
So! A quick note for next time — normally you'd want a disclosure like this to happen privately [via secure channels](https://github.com/keep-network/tbtc/#security), but in this case we discussed first briefly on Discord, and the issue is already on the team's radar and mostly mitigated.

Working through a longer response now :)

**Comment by mhluongo** [comment]
Okay, responding here from the top. The behavior you're seeing is not ideal for the network, but it's also already mitigated in the protocol and only allows signers to inconvenience each other. The default client config should, however, respond better when depositors behave this way rather than relying on the kindness of strangers. For that reason, I believe this issue should be moved to https://github.com/keep-network/keep-ecdsa

A couple misconceptions I'm seeing below.

First, depositors exiting mid-deposit is normal user behavior. While we initially designed an escrowed deposit fee, you can see in the repo history that we determined it was unnecessary because the depositor has to pay the beacon fee, *regardless of whether they fund a deposit* - which acts as a deterrent in a mature network.

The addition of rewards here, however, means stakers are more aggressively looking for ways to undermine each other while the network bootstraps. Each reward interval appears to be a zero-sum game. Unfortunately, most stakers have misunderstood the current rewards mechanism.

> Rewards are paid based on # of successful deposit + redeems

This is not true. I'd recommend you review the `keep-core` repository to see why. However, a number of stakers believe it's true, so seeing this behavior for another month or so makes sense.

My recommended response to this is an existing issue on the client, namely that the default client config should be to call `retrieveSignerPubkey` with some random jitter before the deadline.

We do have some further tweaks to the rewards schedule to make this PvP strategy more obviously suboptimal to an individual staker.. but because we don't want to "change the rules" while people are playing the game, that change is slated for interval 3.

**Comment by mhluongo** [comment]
> From the victims perspective. The only way to fight this exploit today is to increase ETH bond and constantly monitor and notify the abandonend deposits, resulting in higher risk and/or real costs to victims.

This is correct, though if you review the `keep-core` repo and the rewards strategy, you'll see the only time they become victims is if the `retrieveSignerPubkey` is never called
